x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -44,23 +53,68 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
     @Override
     protected PhysicalPlan rule(FilterExec filterExec, LocalPhysicalOptimizerContext ctx) {
         PhysicalPlan plan = filterExec;
-        if (filterExec.child() instanceof EsQueryExec) {
+        if (filterExec.child() instanceof EsQueryExec esQueryExec) {
+            plan = rewrite(filterExec, esQueryExec);
+        } else if (filterExec.child() instanceof EvalExec evalExec && evalExec.child() instanceof EsQueryExec esQueryExec) {
+            plan = rewrite(filterExec, evalExec, esQueryExec);
+        }
+
+        return plan;
+    }
+
+    private FilterExec rewrite(FilterExec filterExec, EsQueryExec esQueryExec) {
+        // Find and rewrite any binary comparisons that involve a distance function and a literal
+        var rewritten = filterExec.condition().transformDown(EsqlBinaryComparison.class, comparison -> {
+            ComparisonType comparisonType = ComparisonType.from(comparison.getFunctionType());
+            if (comparison.left() instanceof StDistance dist && comparison.right().foldable()) {
+                return rewriteComparison(comparison, dist, comparison.right(), comparisonType);
+            } else if (comparison.right() instanceof StDistance dist && comparison.left().foldable()) {
+                return rewriteComparison(comparison, dist, comparison.left(), ComparisonType.invert(comparisonType));
+            }
+            return comparison;
+        });
+        if (rewritten.equals(filterExec.condition()) == false) {
+            return new FilterExec(filterExec.source(), esQueryExec, rewritten);
+        }
+        return filterExec;
+    }
+
+    private Map<NameId, StDistance> getPushableDistances(List<Alias> aliases) {
+        LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
+        aliases.forEach(alias -> {
+            if (alias.child() instanceof StDistance distance && canPushSpatialFunctionToSource(distance, Map.of())) {
+                refs.put(alias.id(), distance);
+            } else if (alias.child() instanceof ReferenceAttribute ref && refs.containsKey(ref.id())) {
+                StDistance distance = refs.get(ref.id());
+                refs.put(alias.id(), distance);
+            }
+        });
+        return refs;
+    }
+
+    private PhysicalPlan rewrite(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
```

Comment:
I was thinking about situations like
```
FilterExec
\_EvalExec
  \_FilterExec
    \_EsSourceExec
```
where both filters could be pushable. We do not account for this here, but would that be a relevant case?

We could get that from
```
FROM idx | WHERE ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) <= 10000 | EVAL dist = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) | WHERE dist > 300
```
or so.

If cases like that are relevant, Costin's comment will become more relevant to expand the situations where the push down works. I.e. rather than looking for a certain pattern, we'll have to look for FilterExecs where substituting reference attributes from previous evals leads to a pushable condition.

But let's keep it focused for now :)

Reply:
I would have thought that consolidating EVAL's and WHERE's into fewer, combined commands (in another rule) could resolve that, but can also imagine that might cause performance issues in some cases (eg. the first WHERE filters down the data so the later EVAL has less work). In any case, I agree that we should investigate and consider cases like this in a later PR. This PR is supposed to tackle performance optimization of a number of key use cases, and now already optimizes more cases than initially envisaged, so I'm happy to push more complex cases like this to later work.

Reply:
I just tried and my example from above gets optimized away in 2 rounds of `PushFiltersToSource`. But the situation from above can be forced with a non-pushable filter, like in
```
FROM idx | WHERE ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) <= 10000 AND field::string == "asdf" | EVAL dist = ST_DISTANCE(field, TO_GEOPOINT("POINT(0 0)")) | WHERE dist > 3000
```

Reply:
I think this scenario could be handled in a followup PR.